### PR TITLE
Add api to allow reading a QgsProject file without loading any layers

### PR DESCRIPTION
**Pull request by nyalldawson**
This causes the project to be read, but no data providers created for
layers (i.e. the same result as loading a project in the app and
choosing to leave bad layers in the project).

It's intended for occasions when you want to efficiently access properties
of a project without the (often very expensive) cost of actually
loading layers. E.g. to just read project metadata, parse the
layer tree structure, extract print layouts or renderer properties, etc.

### CODE REVIEW

**Comment by nyalldawson** [comment]
@elpaso mind looking over this?

File: src/core/qgsvectorlayer.cpp [hunk 1]
```diff
@@ -1436,7 +1436,7 @@ bool QgsVectorLayer::readXml( const QDomNode &layer_node, QgsReadWriteContext &c
   }
 
   QgsDataProvider::ProviderOptions options { context.transformContext() };
-  if ( !setDataProvider( mProviderKey, options ) )
+  if ( ( mReadFlags & QgsMapLayer::FlagDontResolveLayers ) || !setDataProvider( mProviderKey, options ) )
   {
     QgsDebugMsg( QStringLiteral( "Could not set data provider for layer %1" ).arg( publicSource() ) );
```
Comment (elpaso) [inline src/core/qgsvectorlayer.cpp]: Maybe differentiate the debug message for the don't resolve case.

**Comment by rldhont** [comment]
@dmarteau can you take a look ?

File: src/core/qgsproject.h [hunk 2]
```diff
@@ -263,18 +263,33 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
      */
     void clear();
 
+    /**
+     * Flags which control project read behavior.
+     * \since QGIS 3.10
+     */
+    enum ReadFlag
+    {
+      FlagDontResolveLayers = 1 << 1, //!< Don't resolve layer paths. Dramatically improves project read time if the actual data from the layers is not required.
```
Comment (elpaso) [inline src/core/qgsproject.h]: I would make it clearer that when this flag is on the layer won't be actually "loaded".

**Review by elpaso — APPROVED** [review]
LGTM, I just have one concern that is not strictly related to this PR: when testing the bad layer handling a few months ago I found and fixed several code paths where `layer->dataProvider()->something()` was invoked on a nullptr data provider. I wonder if we should double check that the precondition is always verified in all the possible points where such a call is attempted.

**Comment by nyalldawson** [comment]
> I wonder if we should double check that the precondition is always verified in all the possible points where such a call is attempted. 

Yes, agreed. I've fixed a few things here I encountered - specifically related to use of dataProvider()->name() instead of QgsMapLayer::providerType(). But we'll need to continue chipping away and adding more safety here.

We should also consider whether it might be most appropriate to filter out invalid layers by default in QgsMapLayerComboBox and other similar widgets.

File: src/core/qgsproject.h [hunk 3]
```diff
@@ -263,18 +263,33 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
      */
     void clear();
 
+    /**
+     * Flags which control project read behavior.
+     * \since QGIS 3.10
+     */
+    enum ReadFlag
+    {
+      FlagDontResolveLayers = 1 << 1, //!< Don't resolve layer paths (i.e. don't load any layer content). Dramatically improves project read time if the actual data from the layers is not required.
```
Comment (wonder-sk) [inline src/core/qgsproject.h]: @nyalldawson just a nitpick - `1 << 1` is two, maybe better to start from one? :)

Comment (nyalldawson) [reply src/core/qgsproject.h]: Thanks, fixed

**Pull request merged**